substrate/frame/support/src/traits/tokens/fungible/union_of.rs

```diff
@@ -0,0 +1,1005 @@
+			Right(a) =>
+				<Right as fungibles::MutateFreeze<AccountId>>::set_freeze(a, id, who, amount),
+		}
+	}
+	fn extend_freeze(
+		asset: Self::AssetId,
+		id: &Self::Id,
+		who: &AccountId,
+		amount: Self::Balance,
+	) -> DispatchResult {
+		match Criterion::convert(asset) {
+			Left(()) => <Left as fungible::MutateFreeze<AccountId>>::extend_freeze(id, who, amount),
+			Right(a) =>
+				<Right as fungibles::MutateFreeze<AccountId>>::extend_freeze(a, id, who, amount),
+		}
+	}
+	fn thaw(asset: Self::AssetId, id: &Self::Id, who: &AccountId) -> DispatchResult {
+		match Criterion::convert(asset) {
+			Left(()) => <Left as fungible::MutateFreeze<AccountId>>::thaw(id, who),
+			Right(a) => <Right as fungibles::MutateFreeze<AccountId>>::thaw(a, id, who),
+		}
+	}
+}
+
+pub struct ConvertImbalanceDropHandler<
+	Left,
+	Right,
+	Criterion,
+	AssetKind,
+	Balance,
+	AssetId,
+	AccountId,
+>(sp_std::marker::PhantomData<(Left, Right, Criterion, AssetKind, Balance, AssetId, AccountId)>);
+
+impl<
+		Left: fungible::HandleImbalanceDrop<Balance>,
+		Right: fungibles::HandleImbalanceDrop<AssetId, Balance>,
+		Criterion: Convert<AssetKind, Either<(), AssetId>>,
+		AssetKind,
+		Balance,
+		AssetId,
+		AccountId,
+	> fungibles::HandleImbalanceDrop<AssetKind, Balance>
+	for ConvertImbalanceDropHandler<Left, Right, Criterion, AssetKind, Balance, AssetId, AccountId>
+{
+	fn handle(asset: AssetKind, amount: Balance) {
+		match Criterion::convert(asset) {
+			Left(()) => Left::handle(amount),
+			Right(a) => Right::handle(a, amount),
+		}
+	}
+}
+
+impl<
+		Left: fungible::Balanced<AccountId>,
+		Right: fungibles::Balanced<AccountId, Balance = Left::Balance>,
+		Criterion: Convert<AssetKind, Either<(), Right::AssetId>>,
+		AssetKind: AssetId,
+		AccountId,
+	> fungibles::Balanced<AccountId> for UnionOf<Left, Right, Criterion, AssetKind, AccountId>
+{
+	type OnDropDebt = ConvertImbalanceDropHandler<
+		Left::OnDropDebt,
+		Right::OnDropDebt,
+		Criterion,
+		AssetKind,
+		Left::Balance,
+		Right::AssetId,
+		AccountId,
+	>;
+	type OnDropCredit = ConvertImbalanceDropHandler<
+		Left::OnDropCredit,
+		Right::OnDropCredit,
+		Criterion,
+		AssetKind,
+		Left::Balance,
+		Right::AssetId,
+		AccountId,
+	>;
+
+	fn deposit(
+		asset: Self::AssetId,
+		who: &AccountId,
+		value: Self::Balance,
+		precision: Precision,
+	) -> Result<fungibles::Debt<AccountId, Self>, DispatchError> {
+		match Criterion::convert(asset.clone()) {
+			Left(()) => <Left as fungible::Balanced<AccountId>>::deposit(who, value, precision)
+				.map(|inner_debt| {
+					let debt = fungibles::Imbalance::new(asset, inner_debt.peek());
+					fungible::Imbalance::forget(inner_debt);
+					debt
+				}),
+			Right(a) => <Right as fungibles::Balanced<AccountId>>::deposit(
+				a, who, value, precision,
+			)
+			.map(|inner_debt| {
+				let debt = fungibles::Imbalance::new(asset, inner_debt.peek());
+				fungibles::Imbalance::forget(inner_debt);
+				debt
+			}),
```

Comment:
DQ: why can't we return the `inner_debt` here instead of creating a new debt? 

maybe can explain in a comment?

Reply:
Drama Queen? (:
it's two different types. one `Credit<_, UnionOf<...> as fungibles>`, and another `Credit<_, Right as fungibles>`